File: UnknownSize.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * // This is ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface ArrayReader {
 *     public int get(int index) {}
 * }
 */

class Solution {
    public int search(ArrayReader reader, int target) {
        int low = 0;
        int high = 1;
        while(reader.get(high) < target){
            low = high;
            high = high * 2; //Increases the search size by 2 as the size of the array is unknown
        }
        while (low <= high){
            int mid = low + (high - low) / 2; //Calculates middle element
            if (reader.get(mid) == target){ //If the target matches the middle element
                return mid;
            }
            if(target <= reader.get(mid)){
                high = mid - 1; //Moves left
            } else {
                low = mid + 1; //Moves right
            }
        }
        return -1;
    }

}

File: unknown_size.py
# The code implements a search algorithm for an unknown-sized, sorted array using an external interface 'ArrayReader'.
# The search method utilizes binary search to efficiently locate the target value within the array.
# An initial left (l) and right (r) boundary is set, where:
#   - 'l' starts at index 0.
#   - 'r' is obtained through reader.get(-1), which is assumed to return the size or a boundary of the array.
# In each iteration, the mid-point (mid) is calculated to split the search space:
#   - If reader.get(mid) equals the target, mid is returned as the index where the target is found.
#   - If reader.get(mid) is greater than the target, the search shifts to the left half by setting r = mid - 1.
#   - Otherwise, it shifts to the right half by setting l = mid + 1.
# If the target is not found after the loop, -1 is returned, indicating the target does not exist in the array.
# 
# TC: O(log n) - Since binary search reduces the search space by half each iteration.
# SC: O(1) - Constant space is used, as only pointers for the search range and mid-point are maintained.


class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int: # type: ignore
        l, r = 0, reader.get(-1)
        print(r)
        while l <= r:
            mid = (l + r) // 2
            if reader.get(mid) == target:
                return mid
            elif reader.get(mid) > target:
                r = mid - 1
            else:
                l = mid + 1
        return -1
        